### PR TITLE
[ADD] package_hierarchy: Prevent packages in 2 locations at same time 

### PR DESCRIPTION
**Pull request by armandcela**

### CODE REVIEW

File: addons/package_hierarchy/models/stock_picking.py [hunk 1]
```diff
@@ -52,6 +52,20 @@ def action_done(self):
...
+          SELECT DISTINCT UNNEST (ARRAY_AGG(DISTINCT package_id) || 
+            ARRAY_AGG(DISTINCT result_package_id) || ARRAY_AGG(DISTINCT u_result_parent_package_id))
+          FROM stock_move_line AS all_packages WHERE picking_id in %(picking_ids)s
+          ) AS all_unique_packages 
+          WHERE all_unique_packages IS NOT NULL
+          ORDER BY all_unique_packages;
+        """
+        self.env.cr.execute(query, {"picking_ids": tuple(self.ids)},)
+        tuple_unique_packages = self.env.cr.fetchall()
+        packages = [package[0] for package in tuple_unique_packages]
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_picking.py:68]: Suggestion (feel free to resolve): Rename this to `package_ids` to make it extra clear that it isn't a recordset.

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 2]
```diff
@@ -61,9 +61,12 @@ def _compute_children_quant_ids(self):
     @api.depends('quant_ids.package_id', 'quant_ids.location_id', 'quant_ids.company_id', 'quant_ids.owner_id')
     def _compute_package_info(self):
         for package in self:
-            values = {'location_id': False, 'company_id': self.env.user.company_id.id, 'owner_id': False}
-            package._check_not_multi_location()
-            values['location_id'] = package.children_quant_ids.mapped('location_id')
+            values = {
+                'location_id': False, 'company_id': self.env.user.company_id.id, 'owner_id': False
+            }
+            locations = package.children_quant_ids.mapped('location_id')
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:67]: Can we disable prefetching here? Just to be safe.

`locations = package.with_context(prefetch_fields=False).children_quant_ids.mapped('location_id')`

**Review by PeterAlabaster — APPROVED** [review]

**Review by PeterC10 — APPROVED** [review]

File: addons/package_hierarchy/models/stock_quant_package.py (hunk 2 above)
Comment (armandcela) [reply addons/package_hierarchy/models/stock_quant_package.py:67]: I think should be fine

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 3]
```diff
@@ -129,3 +132,19 @@ def action_toggle_processed(self):
...
+        Override method to be able to see package transfers from a parent package too.
+        Default filter will be confirmed picks:
+          state in ('confirmed', 'waiting', 'assigned')
+        """
+        action = self.env.ref("stock.action_picking_tree_all").read()[0]
+        packages = self.search([("id", "child_of", self.ids), ("package_id", "!=", False)])
+        domain = [
+            "|", ("result_package_id", "in", packages.ids), ("package_id", "in", packages.ids)
+        ]
+        pickings = self.env["stock.move.line"].search(domain).mapped("picking_id")
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:147]: Nit: Can we put `["stock.move.line"]` into a variable at the top of the method?

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 4]
```diff
@@ -129,3 +132,19 @@ def action_toggle_processed(self):
...
+    def action_view_picking(self):
+        """
+        Override method to be able to see package transfers from a parent package too.
+        Default filter will be confirmed picks:
+          state in ('confirmed', 'waiting', 'assigned')
+        """
+        action = self.env.ref("stock.action_picking_tree_all").read()[0]
+        packages = self.search([("id", "child_of", self.ids), ("package_id", "!=", False)])
+        domain = [
+            "|", ("result_package_id", "in", packages.ids), ("package_id", "in", packages.ids)
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:145]: Q: Do we need to consider `u_result_parent_package_id`? Might be one for product.

File: addons/package_hierarchy/models/stock_quant_package.py (hunk 3 above)
Comment (armandcela) [reply addons/package_hierarchy/models/stock_quant_package.py:147]: It is used only once, and it was in core method too. But yes can do that

File: addons/package_hierarchy/models/stock_quant_package.py (hunk 4 above)
Comment (armandcela) [reply addons/package_hierarchy/models/stock_quant_package.py:145]: When generally there is no issue result_package_id is a child of u_result_parent_package_id.
In the issues we have seen on multiple locations still is the case. Maybe when we were doing the changes in  a between time there was a moment where result_package_id was not a child of u_result_parent_package_id, which still after we fixed the final should be always.
Can add another or condition `("u_result_parent_package_id", "in", self.filtered(lambda p: not p.parent_id).ids)` -- Maybe better a check before if there are and to add with expressionor

Comment (PeterC10) [reply addons/package_hierarchy/models/stock_quant_package.py:145]: I think it's fine tbh.

It isn't possible now from the picking to have parent packages in two locations so if we had a parent package in two locations somehow then we would go to that record directly.

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 5]
```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
...
     def _check_not_multi_location(self):
         if not self.env.context.get("bypass_multi_location_check", False):
-            for package in self:
-                locations = package.children_quant_ids.mapped('location_id')
-                if len(locations) > 1:
-                    raise ValidationError(_('Package cannot be in multiple '
-                                            'locations:\n%s\n%s') % (package.name,
-                                                                    ', '.join( [l.name for l in locations])))
+            packages = self.filtered(lambda p: p.package_id)
+            parent_packages = self.filtered(lambda p: not p.package_id)
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:43]: As we're making these changes around performance, we could just do:

`parent_packages = self - packages`

Although not sure the gain would be noticeable, also makes the code a bit less explicit.

Just thinking out loud. :)

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 6]
```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
...
+            multi_package_locations = False
+            multi_parent_package_locations = False
+            total_packages = []
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:46]: Can we rename this variable? It is for packages in more than one location right?

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 7]
```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
...
+                multi_package_locations = self.env.cr.fetchall()
+            if multi_parent_package_locations:
+                total_packages += [result[1] for result in multi_parent_package_locations]
+            if multi_package_locations:
+                total_packages += [result[1] for result in multi_package_locations]
+            if total_packages:
+                all_multi_location_packages = self.browse(total_packages)
+                raise ValidationError(
+                    _("Package cannot be in multiple locations:\n%s") %
+                    (", ".join([p.name for p in all_multi_location_packages]))
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:81]: The error message no longer shows the locations where the packages are in.

Might be worth listing each package, and then all the locations it is in, so the issue is easier to resolve.

Comment (PeterC10) [reply addons/package_hierarchy/models/stock_quant_package.py:81]: Actually nevermind, as we discussed the user can now go to the package to find this info. But can we reword the error a bit. E.g.:

```
The following packages cannot be set into multiple locations:

xxx
```

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 8]
```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
...
     def _check_not_multi_location(self):
         if not self.env.context.get("bypass_multi_location_check", False):
-            for package in self:
-                locations = package.children_quant_ids.mapped('location_id')
-                if len(locations) > 1:
-                    raise ValidationError(_('Package cannot be in multiple '
-                                            'locations:\n%s\n%s') % (package.name,
-                                                                    ', '.join( [l.name for l in locations])))
+            packages = self.filtered(lambda p: p.package_id)
+            parent_packages =self - packages
```
Comment (PeterC10) [inline addons/package_hierarchy/models/stock_quant_package.py:43]: Nit: Missing space between `=` and `self`.

Will raise an issue for formatting the repo.

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 9]
```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
...
+                    WHERE sq.package_id in %(package_ids)s
+                    GROUP BY sq.package_id
+                    HAVING COUNT(Distinct sq.location_id) > 1;
+                """
+                self.env.cr.execute(
+                    package_query, {"package_ids": tuple(packages.ids)},
+                )
+                multi_package_locations = self.env.cr.fetchall()
+            if multi_parent_package_locations:
+                total_packages += [result[1] for result in multi_parent_package_locations]
```
Comment (ws-mps) [inline addons/package_hierarchy/models/stock_quant_package.py]: Do we want to use `append()` instead `+=` ? (same for line 76)

Comment (ws-mps) [reply addons/package_hierarchy/models/stock_quant_package.py]: I meant extend :facepalm:

**Pull request merged**